app/plugin_api_test.go

```diff
@@ -659,3 +660,80 @@ func TestBasicAPIPlugins(t *testing.T) {
 		}
 	}
 }
+
+func TestPluginAPIKVCompareAndSet(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+	api := th.SetupPluginAPI()
+
+	testCases := []struct {
+		Description   string
+		ExpectedValue []byte
+	}{
+		{
+			Description:   "Testing non-nil, non-empty value",
+			ExpectedValue: []byte("value1"),
+		},
+		{
+			Description:   "Testing empty value",
+			ExpectedValue: []byte(""),
+		},
+	}
+
+	for i, testCase := range testCases {
+		t.Run(testCase.Description, func(t *testing.T) {
+			expectedKey := fmt.Sprintf("Key%d", i)
+			expectedValueEmpty := []byte("")
+			expectedValue1 := testCase.ExpectedValue
+			expectedValue2 := []byte("value2")
+			expectedValue3 := []byte("value3")
+
+			// Make sure no key is already created
+			value, err := api.KVGet(expectedKey)
+			require.Nil(t, err)
+			require.Nil(t, value)
+
+			require.Nil(t, api.KVSet(expectedKey, expectedValue1))
+			value, err = api.KVGet(expectedKey)
+			require.Nil(t, err)
+			require.Equal(t, expectedValue1, value)
+
+			// Update using correct old value
+			require.Nil(t, api.KVCompareAndSet(expectedKey, expectedValue1, expectedValue2))
+
+			value, err = api.KVGet(expectedKey)
+			require.Nil(t, err)
+			require.Equal(t, expectedValue2, value)
+
+			// Update using incorrect old value
+			err = api.KVCompareAndSet(expectedKey, []byte("incorrect"), expectedValue3)
+			require.NotNil(t, err)
+			require.Contains(t, err.Error(), "no rows were affected")
+			require.Equal(t, err.StatusCode, http.StatusNotModified)
+
+			value, err = api.KVGet(expectedKey)
+			require.Nil(t, err)
+			require.Equal(t, expectedValue2, value)
+
+			// Update using nil old value
+			err = api.KVCompareAndSet(expectedKey, nil, expectedValue3)
+			require.NotNil(t, err)
+			require.Contains(t, err.Error(), "no rows were affected")
+			require.Equal(t, err.StatusCode, http.StatusNotModified)
+
+			value, err = api.KVGet(expectedKey)
+			require.Nil(t, err)
+			require.Equal(t, expectedValue2, value)
+
+			// Update using empty old value
+			err = api.KVCompareAndSet(expectedKey, expectedValueEmpty, expectedValue3)
+			require.NotNil(t, err)
+			require.Contains(t, err.Error(), "no rows were affected")
+			require.Equal(t, err.StatusCode, http.StatusNotModified)
```

Comment:
I wonder if the plugin API client might be better off with a (bool, err) signature vs. having to introspect the `http.StatusNotModified`. This might be more idiomatic, ala the casting check and such.

Reply:
So you proposal is to return:
- `false, err` if some DB error occurred
- `false, nil` if `current value != oldValue`
- `true, nil` if `current value == oldValue`

Is this correct? If so, I like the idea :wink:

Reply:
Ah I see. Yep bubbling it up all the way through to the api will be convenient for the clients.